### DISCUSSION TITLE
Bug? git from Homebrew ignores local timezone in git commit on Linux

### DISCUSSION THREAD

**Discussion by tueda** [Homebrew/discussions#3617] [Linux]
I am struggling with `git` installed by Homebrew on Linux (WSL, Ubuntu 20.04), after executing `brew upgrade` without thinking. Somehow the new bottled `git` ignores my timezone in `git commit` and uses GMT (`+0000`). I also tried building from source (`brew install git --build-from-source`) but the result was the same.

For testing independent of my environment, I created a workflow with GitHub Actions ([repo](https://github.com/tueda/test-bottled-git-timezone)). It basically does the following:
1. Make a commit with the system-default `git`.
2. Install `git` with `brew install git`.
3. Make a commit with the `git` from Homebrew.

The [log](https://github.com/tueda/test-bottled-git-timezone/runs/8226417214) shows (`Run git log --pretty=fuller` at the end)
```
commit e781553f5211b613cdcbfbdc3f3189a5941922e8
Author:     Your Name <you@example.com>
AuthorDate: Wed Sep 7 10:59:40 2022 +0000
Commit:     Your Name <you@example.com>
CommitDate: Wed Sep 7 10:59:40 2022 +0000

    commit with bottled git version 2.37.3

commit 8adf58d68594a0531b7165f4b95585819df29f09
Author:     Your Name <you@example.com>
AuthorDate: Wed Sep 7 19:56:11 2022 +0900
Commit:     Your Name <you@example.com>
CommitDate: Wed Sep 7 19:56:11 2022 +0900

    commit with git version 2.37.3
```
As you see, the `git` coming from Homebrew ignores the timezone (`+0900`, which I set in the workflow) and uses GMT (`+0000`). On the other hand, the same workflow on macOS (actually, the default `git` seems to be the brewed one) shows `git` uses the correct timezone `+0900`. So, it sounds like there is some problem in the build process of `git` on Linux??

**Comment by carlocab** [comment]
> So, it sounds like there is some problem in the build process of `git` on Linux??

The build process is almost exactly identical between Linux and macOS (the only difference is Linux uses OpenSSL and macOS uses CommonCrypto), so that seems unlikely.

Have you tried testing Git 2.37.3 on Linux built without using Homebrew?

**Reply by carlocab** [reply]
Ah, I see the Ubuntu runner already has Git 2.37.3 installed at `/usr/bin/git`. We should compare how Ubuntu configures their version of Git and see how it differs from ours.

**Reply by carlocab** [reply]
Oh, wait. That runs on `ubuntu-latest`, which is Ubuntu 20.04. But on Ubuntu 20.04, [Git is 2.25.1](https://packages.ubuntu.com/source/focal/git). So I'm not sure where the runner is getting a `/usr/bin/git` that has version 2.37.3.

**Reply by carlocab** [reply]
The runner installs git using this script: https://github.com/actions/runner-images/blob/aa116c7fab0293bf99c62405c212fdf5b2a5b928/images/linux/scripts/installers/git.sh

I guess the PPA they use might reveal something about the difference.

**Reply by tueda** [reply]
Thanks for your reply.

Indeed, they say Git 2.37.2 in their image is from apt source repository: ppa:git-core/ppa, in [README](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md).

> Have you tried testing Git 2.37.3 on Linux built without using Homebrew?

Not yet.

**Reply by tueda** [reply]
I tried building and testing Git 2.37.3 without Homebrew on Ubuntu 20.04 by using Docker, and it worked correctly.
<details>
<summary>Details</summary>

```bash
docker run -it --rm ubuntu:focal-20220826
```
```bash
# Install everything we need.
apt-get update
DEBIAN_FRONTEND=noninteractive apt-get install -y build-essential git dh-autoreconf libcurl4-gnutls-dev libexpat1-dev gettext libz-dev libssl-dev install-info

# Set the local timezone.
ln -sf /usr/share/zoneinfo/Asia/Tokyo /etc/localtime

# Git configuration.
git config --global user.email "you@example.com"
git config --global user.name "Your Name"

# Print information.
date
type git       # /usr/bin/git
git --version  # git version 2.25.1

# Make a Git repository and create the first commit.
mkdir test
cd test
git init
git commit --allow-empty -m "first commit"

# Build and install Git from the source.
curl -OL https://mirrors.edge.kernel.org/pub/software/scm/git/git-2.37.3.tar.xz
tar xvf git-2.37.3.tar.xz
cd git-2.37.3
make configure
./configure
make all
make install

# Clear the hash table for command paths.
hash -r

# Print information.
date
type git       # /usr/local/bin/git
git --version  # git version 2.37.3

# Create the second commit.
git commit --allow-empty -m "second commit"

# Show the Git log.
git log
```
```
commit 378ca4c1a0310c835bfa32f823f4f8ceef598ef4 (HEAD -> master)
Author: Your Name <you@example.com>
Date:   Wed Sep 14 19:23:29 2022 +0900

    second commit

commit 687059f4b17d725c4e8e4b3686f4cf13ed7e6852
Author: Your Name <you@example.com>
Date:   Wed Sep 14 19:11:33 2022 +0900

    first commit
```
</details>

For reference, here is testing the brewed Git with the Homebrew Docker image (Ubuntu 20.04), which expectedly fails:

<details>
<summary>Details</summary>

```bash
docker run -it --rm ghcr.io/homebrew/ubuntu20.04:3.6.1 
```
```bash
# Set the local timezone.
sudo ln -sf /usr/share/zoneinfo/Asia/Tokyo /etc/localtime

# Git configuration.
git config --global user.email "you@example.com"
git config --global user.name "Your Name"

# Print information.
date
type git       # git is /usr/bin/git
git --version  # git version 2.37.3

# Make a Git repository and create the first commit.
mkdir test
cd test
git init
git commit --allow-empty -m "first commit"

# Install Git by Homebrew
brew install git

# Clear the hash table for command paths.
hash -r

# Print information.
date
type git       # /home/linuxbrew/.linuxbrew/bin/git
git --version  # git version 2.37.3

# Create the second commit.
git commit --allow-empty -m "second commit"

# Show the Git log.
git log
```
```
commit bc6cb15ec1ba53c0b71e7769faa25250c931fbde (HEAD -> master)
Author: Your Name <you@example.com>
Date:   Wed Sep 14 10:43:50 2022 +0000

    second commit

commit abb50932fdfa2b5d9930bb49fa721a6a43b37196
Author: Your Name <you@example.com>
Date:   Wed Sep 14 19:36:56 2022 +0900

    first commit
```
During `brew install git`, I saw the following warning:
```
==> Installing git dependency: glibc
==> Pouring glibc--2.35.x86_64_linux.bottle.tar.gz
==> Installing locale data for en_US.UTF-8
==> /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35/bin/localedef -i en_US -f UTF-8 en_US.UTF-8
Last 15 lines from /home/linuxbrew/.cache/Homebrew/Logs/glibc/post_install.01.localedef:
2022-09-14 19:38:50 +0900

/home/linuxbrew/.linuxbrew/Cellar/glibc/2.35/bin/localedef
-i
en_US
-f
UTF-8
en_US.UTF-8

failed to set locale!
[error] character map file `UTF-8' not found: No such file or directory
failed to set locale!
[error] default character map file `ANSI_X3.4-1968' not found: No such file or directory
Warning: The post-install step did not complete successfully
You can try again using:
  brew postinstall glibc
```
I guess this is because it was in the Docker image and `brew postinstall glibc` should work for normal usage.
</details>

**Comment by cho-m** [comment]
It could be due to `postinstall` step in `glibc` that adds timezone info. https://github.com/Homebrew/homebrew-core/blob/master/Formula/glibc.rb#L182-L190
```rb
    # Set the local time zone
    sys_localtime = Pathname("/etc/localtime")
    brew_localtime = prefix/"etc/localtime"
    (prefix/"etc").install_symlink sys_localtime if sys_localtime.exist? && brew_localtime.exist?

    # Set zoneinfo correctly using the system installed zoneinfo
    sys_zoneinfo = Pathname("/usr/share/zoneinfo")
    brew_zoneinfo = share/"zoneinfo"
    share.install_symlink sys_zoneinfo if sys_zoneinfo.exist? && !brew_zoneinfo.exist?
```

Looks a bit suspicious that the conditional differs between two parts. Maybe `brew_localtime.exist?` should be negated (i.e. `!brew_localtime.exist?`)

---

Can try testing that by modifying formula (related diff below) and then running `brew postinstall glibc`,
```diff
diff --git a/Formula/glibc.rb b/Formula/glibc.rb
index 7586de2c591..ec933d118a7 100644
--- a/Formula/glibc.rb
+++ b/Formula/glibc.rb
@@ -182,7 +182,7 @@ class Glibc < Formula
     # Set the local time zone
     sys_localtime = Pathname("/etc/localtime")
     brew_localtime = prefix/"etc/localtime"
-    (prefix/"etc").install_symlink sys_localtime if sys_localtime.exist? && brew_localtime.exist?
+    (prefix/"etc").install_symlink sys_localtime if sys_localtime.exist? && !brew_localtime.exist?

     # Set zoneinfo correctly using the system installed zoneinfo
     sys_zoneinfo = Pathname("/usr/share/zoneinfo")
```

Or manually doing the symlink creation (not exactly the same as this will be absolute path as opposed to relative path by Homebrew)
```
ln -s /etc/localtime /home/linuxbrew/.linuxbrew/Cellar/glibc/2.35/etc/localtime
```
